cmd/fs_handlers.go

```diff
@@ -47,39 +46,43 @@ func handleStage(ctx *cli.Context, ctl *client.Client) error {
 		return fmt.Errorf("Failed to retrieve absolute path: %v", err)
 	}
 
-	info, err := os.Stat(absLocalPath)
+	_, err = os.Stat(absLocalPath)
 	if err != nil {
 		return err
 	}
 
-	if info.IsDir() {
-		return handleStageDirectory(ctx, ctl, absLocalPath, repoPath)
-	}
-
-	if !info.Mode().IsRegular() {
-		return fmt.Errorf("not adding non-regular file: %s", absLocalPath)
-	}
+	return handleStageDirectory(ctx, ctl, absLocalPath, repoPath)
+}
 
-	return ctl.Stage(absLocalPath, repoPath)
+type stagePair struct {
+	local, repo string
 }
 
-func handleStageDirectory(ctx *cli.Context, ctl *client.Client, root, repoRoot string) error {
-	// First create all directories:
-	type stagePair struct {
-		local, repo string
+func walk(root, repoRoot string, depth int) ([]stagePair, error) {
+	toBeStaged := make([]stagePair, 0)
+	toBeDereferenced := make([]stagePair, 0)
+	depth++
+	if depth > 255 {
+		return toBeStaged, fmt.Errorf("Exceeded allowed dereferencing depth")
 	}
-
-	toBeStaged := []stagePair{}
-
 	root = filepath.Clean(root)
 	repoRoot = filepath.Clean(repoRoot)
-
 	err := filepath.Walk(root, func(childPath string, info os.FileInfo, err error) error {
 		repoPath := filepath.Join("/", repoRoot, childPath[len(root):])
 
-		if info.IsDir() {
-			if err := ctl.Mkdir(repoPath, true); err != nil {
-				return e.Wrapf(err, "mkdir: %s", repoPath)
+		if info.Mode()&os.ModeSymlink != 0 {
+			resolvedPath, err := filepath.EvalSymlinks(childPath)
+			if err != nil {
+				return fmt.Errorf("Failed to resolve: %v: %v", childPath, err)
+			}
+			info, err = os.Stat(resolvedPath)
+			if err != nil {
+				return fmt.Errorf("Failed to do os.Stat(%v): %v", resolvedPath, err)
+			}
+			childPath = resolvedPath
+			if info.Mode().IsDir() {
+				toBeDereferenced = append(toBeDereferenced, stagePair{childPath, repoPath})
```

Comment:
More a musing: I'm a little confused by the usage of this ``toBeDereferenced`` slice. What would stop you from calling ``walk()`` again directly?

----

Also after some more thinking I found another issue: This method can lead to adding the same file more than once. I mentioned earlier that symbolic links transform a file tree into a graph. This leads to the possibility of encountering the same node more than once while traversing (i.e. directory symlink points one directory level up: ``..``).

So maybe it's actually worth looking at libraries like [symwalk](https://pkg.go.dev/github.com/facebookgo/symwalk) which claim to handle this stdlib weakness. Just this disclaimer is a bit embarassing:

```
CAVEAT: Note that symwalk.Walk does not terminate if there are any non-terminating loops in the file structure.
```

Maybe there are better libraries also :)

Reply:
I looked at the `symwalk` code and that CAVEAT inspired me to write our own,
it is not very different, from our walk. The main change was to track `depth` of recursion. Honestly, speaking I would say having that `walk` in the library is somewhat overkill. It is a typical undergraduate problem, with 10 LOC solution.

Admittedly, I made it a bit more complicated. As you suggest, instead of collecting `toBeDereferenced`, I could have jump to the `walk` right away. It would be cleaner, and I will redo it accordingly.

Traveling same node more than once is a *feature* of the symlinks. So this is a necessary evil. I even tested  our code `dir1/dir2 --> ../dir1`, it creates quite a long unwraped tree (limited by the set depth).